[1단계 - 로또(자동)] 마루(송채원) 미션 제출합니다.

chawani opened the pull request:
토니 안녕하세요😁 마루입니다!
로또 미션을 진행해봤는데, 지난 미션보다 더 복잡하고 어려워진거 같아요..!

궁금한점 두가지가 생겨서 질문 드립니다!

1. 원시값을 구분하는 명확한 기준이 무엇일까요? 모든 변수에 대해서 원시값 포장을 해야하는 걸까요?
2. 생성자를 만들때, 최대한 많은 타입의 파라미터를 받을 수 있도록 생성자를 여러개 생성하는건 좋은건가요? 생성자를 여러개 만들때 괜찮은 기준이 어디까지인지 잘 모르겠어요😥 그리고 생성자로 전달받는 데이터의 타입은 클래스가 가지고 있는 주요 속성의 타입과 일치하는것이 좋은가요?

리뷰 잘 부탁드립니다! 감사합니다..!

kyukong committed:
docs: 구현해야 하는 기능 작성

kyukong committed:
feat: 구입 금액 객체 생성

kyukong committed:
feat: 숫자 객체 생성

kyukong committed:
refactor(PurchaseAmount): 파라미터 변경 불가하도록 수정

kyukong committed:
feat: 로또 객체 생성

kyukong committed:
feat: 1에서 45 사이의 랜덤 값 가지는 Ball 객체 생성

kyukong committed:
feat: 입력받은 개수만큼 로또 객체들 생성

chawani committed:
fix(PurchaseAmount): 검증하는 로직에서 판단 오류 수정

chawani committed:
refactor: 코드 리팩터링

- 사용하지 않는 import 삭제
- 사용하지 않는 Ball 생성자 삭제
- Lotto 매직넘버 상수화
- 테스트코드 스네이크 케이스로 수정

chawani committed:
feat: 입력 금액에 맞는 로또 개수 반환 구현

chawani committed:
feat: 구입금액을 입력받고 로또들을 구매하는 기능 구현

chawani committed:
feat: 지난 주 당첨 번호 입력받는 기능 구현

chawani committed:
feat: 로또 결과 객체 생성

- Rank 열거형 생성

kyukong committed:
feat: 로또 일치 결과 계산하여 출력 및 Rank 클래스 속성 추가

- 간결한 네이밍 수정
- 매직넘버 상수화
- 테스트코드 간결화

kyukong committed:
feat: 수익률 계산하여 출력

kyukong committed:
refactor: 지난 주 당첨번호와 보너스 볼을 가지는 WinnigLotto 객체 생성

kyukong committed:
refactor: 입력 에러 발생 시 재입력 기능 추가

- Output 에서 에러 메시지 출력
- 사용하지 않는 util 클래스 삭제

kyukong committed:
fix: 숫자 판단 정규식 수정

- 음수도 포함

toneyparky commented on src/main/java/lotto/domain/Profit.java:
```diff
@@ -0,0 +1,12 @@
+package lotto.domain;
+
+public class Profit {
```
마루의 첫 번째 질문에 대해서 이야기해볼까요?

원시값과 도메인 객체에 대한 질문이군요. 

마루가 실제로 개발을 해보며 느낀 점을 들어보고 싶어요. 왜 수익이라는 개념을 포장했고 이를 통해 어떤 이점이 있었는지 궁금합니다!

chawani replied:
수익이라는 개념을 포장하면서 수익률 계산에 대한 역할을 다른 클래스로부터 분리할 수 있었습니다! 그리고 로또 숫자를 저장하는 클래스를 만들었을때 로또 숫자 범위에 대한 유효성 검사를 할 수 있었습니다. 
어떤 속성에 대한 동작이 필요할때, 유효성 검사가 필요할 때 일급컬렉션을 만드는 것이라고 이해하면 될까요??

toneyparky replied:
일반적으로 원시 값을 객체로 포장할 때에는 원시 값이 단지 값이 아니라 프로그램을 위한 특정한 행위를 수행하고 책임을 가질 때에요.

속성에 대한 동작이 필요할 때, 유효성 검사가 필요할 때는 원시 값이 어떤 행위를 하거나 원시 값을 생성할 때에 지켜야하는 책임이 있는 경우네요. 마루가 잘 파악해주신거 같아요 👍

말씀해주셨듯, 수익이라는 개념이 원시값으로 구현될 경우 원시 값 수익이 포함된 클래스에서 수익을 계산하는 행위를 수행하겠죠. 하지만 마루는 수익을 계산하는 행위를 원시값이 담당하도록 처리했어요.

또한 로또 숫자를 저장하는 클래스가 자체적으로 유효성 검사를 함으로써 클래스 자체가 자신이 가지는 숫자의 범위를 제한하는 책임을 가지게 되어요.

모든 동작에 대해, 모든 유효성 검사에 대해서 객체로 만들어야한다고 생각하지 않아요. 1+1은 2다와 같이 명확한 기준이 아니기에 어려운 부분이에요. 이는 경험을 통해 구분하는 능력을 기를 수 있다고 생각해요. 저도 기르고 있는 중이구요. 다만 객체를 만들 때에는 꼭 책임과 행위에 대해서 고민해야해요. 프로그램을 설계하면서 책임과 행위를 먼저 뽑아낸 뒤, 이 책임과 행위를 적절하게 수행할 객체를 만들어내는 방법도 있답니다.

이 부분에 대한 깊은 이야기는 조영호님의 서적을 읽어보시면 좋아요. 우테코 1 단계 필독도서인 객체지향의 사실과 오해 (http://www.yes24.com/Product/Goods/18249021) 아직 읽어보지 않으셨다면 추천드립니다!

마지막으로 한가지 더!

일급 컬렉션을 언급해주셨기에 아래의 글을 소개드릴게요! 글을 참고하여 학습해보면 좋을거에요 😀 :
https://tecoble.techcourse.co.kr/post/2020-05-08-First-Class-Collection/

toneyparky commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,33 @@
+
+    public Lottos(List<Lotto> lottos) {
+        this.lottos = new ArrayList<>(lottos);
+    }
```
이번엔 두 번째 질문에 대해서 이야기해볼게요! 

지금 코드에서 마루는 두가지 생성자를 만들었네요! 이렇게 다른 인자를 가지는 생성자를 만들었을 때에 느꼈던 장점이 있을까요? 

그리고 마루가 이야기하신대로 전달 받는 데이터의 타입이 클래스의 필드와 동일한 `List<Lotto>`인 생성자만 있을 때에는 어떤 방식을 이용하여 `int count`로 `Lottos` 객체를 생성할 수 있을까요?

두가지 질문입니다 😊

chawani replied:
1. 다른 인자를 가지는 생성자들을 만들면 다양한 입력에 대응해서 객체를 만들 수 있다는 장점이 있는거같습니다. 생성자가 다른 생성자를 호출하면서 재사용도 가능하구요.. 하지만 이렇게 생각하면 생성자를 일어날 수 있는 모든 입력에 대해 만들어둬야 할 거같은 느낌이 있는데, 어디까지가 적정선인지 궁금합니다..!
2. 1번을 적고 생각해보니 정적 팩토리 메소드를 사용하면 이름을 가지고 객체를 생성할 수 있기때문에 생성자가 여러개가 된다는 껄끄러움이 없어질 수 있을 거 같습니다! 그러면 count로 Lottos 객체를 생성할 때는 정적 팩토리 메소드를 사용해서 로또 리스트를 만드는 로직을 수행하고 List<Lotto>를 필드로 받는 생성자에 넘겨주면 될거같습니다! 괜찮은 방법인가요?

toneyparky replied:
1. 마루가 말씀하신 다양한 생성자의 장점에 저도 공감해요 👍 하지만 어디까지 생성자를 만들고 만들지 말아야 한다는 규칙은 없는걸로 알고 있습니다. 그때 그때 상황에 맞게 필요한 생성자가 있다면 만들어주면 되어요! 
단, 지금 사용하지 않는 생성자를 미리 만드는 건 지양해야합니다! 현 시점에서 불필요한 코드니까요!

마루가 하는 고민을 해결하기 위해서 빌더 패턴을 사용하기도 해요. 아래의 글을 읽고 지금의 구현과 빌더 패턴의 차이를 정리해주시면 학습에 도움이 될거에요!
https://bbaktaeho-95.tistory.com/72


2. 적어주신대로 정적 팩터리 메서드는 이름을 통해 명확함을 줄 수 있다는 특징이 있죠! 
실제로 마루가 말씀해주신 방식으로 정적 팩터리 메서드를 사용하기도 하구요.
지금의 구현도 좋고 정적 팩터리 메서드를 사용하시려는 구현도 좋네요!

저는 마루는 생성자가 여러개가 되었을 때에 왜 껄끄러움을 느끼셨는지 궁금해요 😬

chawani replied:
만약에 로또라는 클래스에 생성자를 만들때
1. 로또 번호를 Integer 리스트로 받는 생성자
2. 로또 번호를 String 리스트로 받는 생성자
3. 로또 번호를 LottoNumber라는 객체 리스트로 받는 생성자

등등 여러가지 경우가 나올 수도 있는데 로또를 생성할때마다 해당 로직에 편리한대로 생성자를 만들어두다보면 외부에서 객체를 생성할 때 통일성을 헤친다는 느낌이 들어서 약간 껄끄럽다고 느껴지는거같아요.. 별로 신경쓰지 않아도 될 문제일까요?😂

toneyparky replied:
외부에서 객체를 생성할 때의 통일성이라는 부분을 고민하셨군용

크게 고민하지 않으셔도 되는 문제라고 생각해요! 오히려 다양한 생성자를 지원한다면 해당 객체가 유연하게 사용될 수 있다는 장점이 있죠! 다만 생성자가 너무 많아지면 코드를 읽는 데에 어려움을 느낄 수 있겠지만요!

toneyparky commented on README.md:
```diff
@@ -5,3 +5,34 @@
 
 - [온라인 코드 리뷰 과정](https://github.com/woowacourse/woowacourse-docs/blob/master/maincourse/README.md)
+
+## 🎱 구현해야 하는 기능
```
크흐.. 깔끔한 예외 정리네요! 

todo 를 활용하여 체크해주신 부분이 인상깊어요! 👍

다만 적어주신 `Illegalargumentexception`는 주로 `IllegalArgumentException`라는 이름의 클래스를 언급하기에 표준 스타일을 맞춰주면 좋겠죠?

toneyparky commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,97 @@
+
+        Lottos lottos = buyLotto(purchaseAmount);
+
+        createLottoResult(lottos, purchaseAmount);
```
위의 `PurchaseAmount`를 반환하는 `createPurchaseAmount`를 보고 아래 `createLottoResult` 함수를 본다면 이 친구도 뭔가를 반환해야할 것 같은 느낌이 들지 않나요?

코드를 읽는 사람은 앞선 맥락을 기억해요. 이를 잘 고려해서 이해하기 쉽도록 수정해볼까요?

toneyparky commented on src/main/java/lotto/view/Input.java:
```diff
@@ -0,0 +1,23 @@
+
+    public static String bonusBall() {
+        return input();
+    }
```
함수는 보통 동사로 시작하죠! 물론 정적인 유틸성 클래스의 메서드기에 지금과 같이 구현해도 `Input.winNumber()`와 같이 클래스 이름과 함께 이해될 수있어요. 

그렇지만 명사보다는 동사형을 사용하는편이 일반적인 컨벤션이에요 😁

toneyparky commented on src/main/java/lotto/domain/Ball.java:
```diff
@@ -0,0 +1,49 @@
+        this.number = Integer.parseInt(number);
+    }
+
+    private void checkValidValue(final String value) {
```
요런 경우에 저는 주로 validate라는 동사를 사용합니다. 어휘를 풍부하게 만들어드릴 조언이라 생각해주세요 😬

```suggestion
    private void validateNumber(final String value) {
```

toneyparky commented on src/main/java/lotto/domain/Ball.java:
```diff
@@ -0,0 +1,49 @@
+
+    private boolean isNumber(final String value) {
+        return value.matches(NUMBER_MATCHES);
+    }
```
요 부분도 작은 팁이고 사람마다 취향이 다를 수 있어요.  

```suggestion
        if (isBlank(value) || isNotNumber(value)) {
            throw new IllegalArgumentException(ERROR_ONLY_NUMBER);
        }
        int number = Integer.parseInt(value);
        checkLottoNumber(number);
    }

    private boolean isBlank(final String number) {
        return number == null || number.isEmpty();
    }

    private boolean isNotNumber(final String value) {
        return !value.matches(NUMBER_MATCHES);
    }
```

위와 같이 메서드를 사용할 때에 !를 붙여서 부정임을 표현하는가 Not을 붙여서 표현하는가에 대한 차이랄까요?!

마루는 어떤쪽이 좋으신가요? 그리고 그 이유는요?

chawani replied:
저도 보통은 isNotNumber로 쓰는 편인데, 제 페어분께서 만약에 숫자인지를 체크해야한다면 !isNotNumber가 되니까 이중부정이 뭔가 어색할 수 있다고 하시더고요..! 생각해보니 진짜 그게 어색한 느낌이 들어서 저도 동의하고 isNumber로 고쳤었는데, 다시 생각해보니 옆에 있는 isBlank와 비슷한 흐름을 가져가려면 isNotNumber를 써야할것도 같아서ㅋㅋㅋ 어떻게 쓰는게 조금 더 좋은 방법일까요..?!

toneyparky replied:
이 질문을 드린 이유는 취향이 갈리기 때문이에요 😇

이처럼 어느 한 쪽을 선택해야 하는 확실한 근거가 없는 문제에 대해서는 같이 협업하는 사람들과 충분한 시간을 가지고 토의한 후에 결정하는 편이 좋답니다! 마루와 페어는 그렇게 고민하고 결정해주신거 같아서 좋아요!

마루가 질문 주신 어떤 게 더 좋은 방법인지에 대해서는 저도 잘 모르겠다는 답변을 드릴게요 😉

(제 취향은 `isNotNumber`를 만들고 만일 `isNumber`가 필요한 때가 오면 따로 `isNumber` 메서드도 만들어줄 거 같아요)

toneyparky commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,85 @@
+
+    public int getMatchingCount(Lotto compareLotto) {
+        return (int) lotto.stream()
+                .filter(compareLotto::contains).count();
```
요런 경우에는 가독성을 위해 . 뒤를 한 칸 내려주면 좋아요 😀
```suggestion
                .filter(compareLotto::contains)
                .count();
```

toneyparky commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,85 @@
+        if (distinctNumbers.size() != numbers.size()) {
+            throw new IllegalArgumentException(ERROR_DUPLICATED_NUMBER);
+        }
+    }
```
코드를 읽는 사람 입장에서 `checkValidNumbers`를 읽으며 내부에 `checkDuplicatedNumber`를 먼저 읽겠죠~

그러나 밖에는 `checkLottoCount`가 먼저 있네요! 

앞서 컨트롤러에서 읽는 이의 맥락에 대해서 이야기했죠? 

여기에도 맥락의 개념이 적용될거 같아요! 

사용자가 부드럽게 읽을 수 있도록 만들어볼까요?

toneyparky commented on src/main/java/lotto/domain/Ball.java:
```diff
@@ -0,0 +1,49 @@
+    private static final String ERROR_ONLY_NUMBER = "숫자를 입력해주세요!";
+    private static final String ERROR_LOTTO_NUMBER = "로또 숫자가 아닙니다";
+
+    private final int number;
```
int와 Integer의 차이에 대해서 알고 계시나요~?

댓글에 남겨주세요 😊

chawani replied:
Integer는 int 자료형을 클래스화 시킨것이라고만 이해하고 있었습니다..!
더 찾아보니 int는 null값을 처리할 수 없는 반면 Integer는 null값을 처리할 수 있어서 SQL과 연동할 경우에 편리하다고 하네요! 더 알아야 할 포인트 같은 게 있을까요?

toneyparky replied:
핵심을 잘 찾아주셨네요! null이 할당 가능한가, 그렇지 않은가! 좋습니다.

마루가 더 학습하고 싶어하시기에 키워드를 몇 개 던져드릴게요.

- 자바의 primitive type과 boxed primitive type 차이
- Boxed primitive type의 캐시 (cache)

toneyparky commented on src/main/java/lotto/domain/LottoMachine.java:
```diff
@@ -0,0 +1,10 @@
+package lotto.domain;
+
+public class LottoMachine {
```
마루는 이 객체의 책임이 뭐라고 생각하시나요? 

필드를 가지지 않고 입력받은 값을 변환해주는 역할만 담당하는군요?!

chawani replied:
책임이 애매한거같아서 Lottos에 돈으로 초기화하는 생성자를 추가했습니다..!

toneyparky commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,33 @@
+            int matchingCount = lotto.getMatchingCount(winningLotto.getWinningLotto());
+            boolean isBonus = matchingCount == CHECK_BONUS_COUNT && lotto.contains(winningLotto.getBonusBall());
+            lottoResult.increaseRankCount(Rank.getRank(matchingCount, isBonus));
+        }
```
이 부분의 로직이 다소 복잡해보여요. 

객체의 책임을 다시 생각해볼까요? 

힌트는 여러장의 로또를 나타내는 일급 컬렉션인 `Lottos`가 로또가 몇 개 맞아야 보너스 에 해당하는지에 대한 규칙을 알고 있다는 점에서 시작해보죠!

toneyparky commented on src/main/java/lotto/domain/Profit.java:
```diff
@@ -0,0 +1,12 @@
+
+    private double profitRate = 0;
+
+    public double calculate(int totalMoney, PurchaseAmount purchaseAmount) {
```
함수의 이름만 보면 `profit`이 계산되어 반환될 거 같아요. 하지만 실제로는 `profitRate` 네요!

읽는이가 혼란스러울 수 있겠다는 생각이 들어요 😉

toneyparky commented on src/main/java/lotto/domain/PurchaseAmount.java:
```diff
@@ -0,0 +1,49 @@
+        }
+    }
+
+    private void checkDivideLottoPrice(final int value) {
```
이 함수의 이름에 대해서 조금 더 고민해볼까요?

chawani replied:
1000원으로만 나누어 떨어지는 가격을 받는 것은 너무 한정적이라고 생각하여 해당 메서드는 삭제하였습니다..!

toneyparky commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,55 @@
+    SECOND(30000000, 5, true),
+    THIRD(1500000, 5, false),
+    FOURTH(50000, 4, false),
+    FIFTH(5000, 3, false),
```
```suggestion
    FIFTH(5_000, 3, false),
```

언더바를 이용하여 긴 숫자는 가독성을 높힐 수 있답니다 !

toneyparky commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,55 @@
+        this.isBonus = isBonus;
+    }
+
+    public static Rank getRank(int matchingCount, boolean isBonus) {
```
정적 메서드네요~! 그리고 반환값은 Rank 자체구요! 

정적 팩토리 메서드라는 키워드를 이용하여 학습해보고 적용해보셔도 좋겠어요! 찾아보시면서 정적 팩토리 메서드의 네이밍 컨벤션도 함께요 😬

toneyparky commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,55 @@
+    public static Rank getRank(int matchingCount, boolean isBonus) {
+        return Arrays.stream(values())
+                .filter(rank -> rank.count == matchingCount)
+                .filter(rank -> rank.isBonus == isBonus)
```
필터를 두 번 사용하셨네요! 

하나로 줄여볼 수 있을까요?

toneyparky commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,55 @@
+        return this.count;
+    }
+    
+    public boolean getIsBonus() {
```
boolean 값을 반환하는 함수는 보통 is로 시작합니다 !

요 글을 참고해볼까요? 
https://stackoverflow.com/questions/5322648/for-a-boolean-field-what-is-the-naming-convention-for-its-getter-setter

toneyparky commented on src/main/java/lotto/view/Output.java:
```diff
@@ -0,0 +1,79 @@
+    private static final String PROFIT = "이익";
+    private static final String LOSS = "손해";
+
+    public static void error(String errorMessage) {
```
`Input`과 마찬가지로 정적 메서드지만 함수의 이름은 동사로 시작하는게 좋겠죠?

아래의 `askPurchaseAmount`는 동사로 시작하기에 규칙이 일관되지 않구요.

toneyparky commented on src/main/java/lotto/view/Output.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    public static void profitRate(final double profitRate) {
+        double splitProfitRate = Math.floor(profitRate * 100) / 100.0;
```
어떤 의미의 변수인가요? split에 나누다라는 의미를 담고 싶으셨을까요?

toneyparky commented on src/main/java/lotto/view/Output.java:
```diff
@@ -0,0 +1,79 @@
+    public static void profitRate(final double profitRate) {
+        double splitProfitRate = Math.floor(profitRate * 100) / 100.0;
+
+        String profitOrNot = PROFIT;
```
이럴 수도 있고 아닐 수도 있다는 변수의 이름보다 명확하게 분리하여 표현해볼까요?

toneyparky commented on src/test/java/lotto/domain/BallTest.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"0", "46"})
```
요렇게 여러 테스트 케이스를 확인할 수 있는 기능을 활용하신거 좋아요 😊

toneyparky commented on src/test/java/lotto/domain/LottoResultTest.java:
```diff
@@ -0,0 +1,19 @@
+
+        assertEquals(lottoResult.getLottoResult().get(Rank.FIRST), 1);
+    }
+}
```
요 부분에 경고 표시가 나타나있네요! 파일의 가장 마지막 줄이 비어있지 않아서 그래요. 


<img width="575" alt="image" src="https://user-images.githubusercontent.com/37579660/155541486-e502e170-deaf-443c-b2cb-44a76bdf2fc5.png">

깃헙이 이렇게 경고를 띄우는 이유는 뭘까요?

chawani replied:
옛날에 C 컴파일러인 gcc가 한 줄 씩 코드를 읽다보니, 마지막에 EOF가 없으면 한 줄이 끝나지 않았다는 것으로 받아들여서 코드를 다른 의도로 읽어버릴 수도 있고, merge 할때도 오류가 날 수 있다고 하네요! 그러면 이 문제가 모든 컴파일러에서 똑같이 적용이 되어서 경고가 뜨는 것일까요?

toneyparky replied:
마루가 이야기한 내용이 근본적인 원인이구요. 다만 자바의 컴파일러는 해당 문제에 적용되지 않는걸로 알고 있어요. 

보다 실질적인 이유로는 터미널 환경에서 해당 파일을 cat 할 때에 마지막 줄이 공백이 아니라면 출력 후에 커서가 뒤에 바로 붙어서 나오기 때문으로 알고 있어요. 

알아보시느라 수고하셨어요 😀👍

toneyparky commented on src/test/java/lotto/domain/LottoTest.java:
```diff
@@ -0,0 +1,67 @@
+    @Test
+    @DisplayName("로또 생성 시 Ball 숫자 중복")
+    void ball_duplicated() {
+        List<String> numbers = new ArrayList<>(Arrays.asList("1", "2", "2"));
```
로또를 생성하는 유효성 검사 구현의 순서가 아래와 같이 바뀐다면 이 테스트는 어떤 내용을 검증하게 될까요?

```java
    private void checkValidNumbers(final List<String> numbers) {
        checkLottoCount(numbers);
        checkDuplicatedNumber(numbers);
    }
```

toneyparky commented on src/test/java/lotto/domain/LottoTest.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    @Test
+    @DisplayName("로또의 숫자가 모두 일치할때")
```
```suggestion
    @DisplayName("로또의 숫자가 모두 일치할 때")
```

맞춤법을 지켜주면 좋겠죠?

toneyparky commented on src/test/java/lotto/domain/LottosTest.java:
```diff
@@ -0,0 +1,65 @@
+
+        lottos.addMatchingCount(lottoResult, winningLotto);
+
+        assertEquals(lottoResult.getLottoResult().get(Rank.FIRST), 1); // 1
```
`Lottos`에 대한 테스트인데 `LottoResult`를 이용하여 검증하고 있네요~!

앞선 `addMatchingCount` 리뷰와 연관된 부분이라 생각합니다~!

적절한 테스트인지 생각해보죠 😊

toneyparky commented on src/test/java/lotto/domain/ProfitTest.java:
```diff
@@ -0,0 +1,38 @@
+        Profit profit = new Profit();
+
+        PurchaseAmount purchaseAmount = new PurchaseAmount("3000");
+        int totalMoney = lottoResult.getTotalMoney(); // 2030050000
```
불필요한 주석이 남겨져 있네요 !

toneyparky commented on src/test/java/lotto/domain/RankTest.java:
```diff
@@ -0,0 +1,15 @@
+    @Test
+    @DisplayName("숫자 입력 시 랭크 반환")
+    void insert_number_get_rank() {
+        assertEquals(Rank.getRank(4, false), Rank.FOURTH);
```
```suggestion
        assertEquals(Rank.getRank(8, false), Rank.FOURTH);
```

요런 식이라면 어떻게 동작할까요?

toneyparky reviewed (changes requested):
안녕하세요 마루! 반갑습니다. 

로또 미션을 요구사항에 맞게 잘 구현해주셨네요!

객체나 일급 컬렉션으로 잘 포장하신 점이 인상깊었습니다. 이 부분에 대해서도 마루가 질문해주셨는데 코멘트 달아뒀으니 확인 부탁드릴게요 😎

궁금하신 사항이나 질문에 대한 답은 대댓글과 DM으로 부탁드려요!

TMI: 저희 집 강아지 이름이 마루라 더 반가웠어요!

![IMG_6908](https://user-images.githubusercontent.com/37579660/155543616-ed1d5d5e-3b7a-4d4b-a02f-c5461246fc63.JPG)

chawani committed:
refactor: 컨트롤러 코드 리팩터링

- 로또 개수를 계산해주는 클래스를 제거

chawani committed:
refactor: 로또 당첨 등수를 카운트 하는 역할을 LottoResult가 하도록 수정

- 구매한 로또가 몇 등 당첨인지 반환하는 역할은 Lotto가 하도록 수정

chawani committed:
docs: 리드미 수정

- 대소문자 표준스타일에 맞춰서 변경

chawani committed:
refactor: Ball과 Lotto의 생성자를 위주로 리팩터링

- Ball을 LottoNumber로 변경
- LottoNumber와 Lotto의 생성자 파라미터 타입을 변경
- 숫자 유효성 검사를 분리하여 NumberValidator 생성
- 유효성 검사를 책임에 알맞은 곳으로 이동

chawani committed:
refactor: view 패키지의 Input과 Output 리네이밍

- Input을 InputView로 리네이밍
- Output을 OutputView로 리네이밍
- 메서드 이름들이 동사로 시작할 수 있도록 리네이밍

chawani committed:
refactor: Profit 클래스가 알맞은 필드와 역할을 가지도록 변경

- 메서드 리네이밍
- 필드 변경
- 생성자 추가

chawani committed:
refactor: PurchaseAmount를 Money로 변경

- 클래스 리네이밍에 맞춰서 유효성 검사 책임 이동
- 1000원 단위로 나누어 떨어지는지에 대한 검증 삭제

chawani committed:
refactor: 금액을 계산하는 로직을 Money로 이등 및 금액을 속성으로 가지는 클래스들을 수정

- int 자료형으로 저장했던 금액을 Money로 변경
- Rank를 반환하는 메서드를 정적 팩토리 메서드로 변경
- Rank 메서드 리네이밍

chawani committed:
fix: OutputView의 printLottoResult 출력 오류 수정

- Rank의 money 속성을 prizeMoney로 리네이밍

chawani committed:
refactor: OutputView의 모호한 변수 삭제 및 상수 리네이밍

chawani committed:
fix: LottoController의 지불 금액 관련 수정

- 구매한 로또의 총 가격이 아니라 지불하려던 총 금액으로 수익률을 계산하는 오류 수정
- purchaseAmount를 LottoController의 속성에서 삭제

chawani committed:
fix: LottoNumber의 equals에 null 처리 추가 및 테스트 추가

chawani committed:
test: 테스트 케이스 추가 및 코드 정리

chawani committed:
refactor: 외부에서 내부 값을 바꾸지 못하도록 getter 수정

- 메서드 이름 오타 수정

chawani commented:
토니 안녕하세요😎 강아지 마루 정말정말 귀엽네요😍 제가 닉네임 지을 때 반려견 이름이 마루인 사람을 만날 수도 있겠다 싶었는데, 그런 일이 실제로 일어났네요😲!ㅋㅋㅋ

이번에 리팩터링 하면서 신경 쓴 부분은
1. 도메인이 각자 알맞은 책임을 가지고 있는가?
2. 컨트롤러 부분을 명료하게 수정하자!

였는데 제대로 됐을진 모르겠네요..! 리뷰 주신 부분도 하나씩 체크해가면서 고쳐봤는데, 혹시나 빠트린 부분 있으면 알려주세요!

그리고 궁금한 점이 한 가지 더 생겼습니다😢 기존에는 도메인들이 생성자 파라미터로 대부분 String을 받고 있었는데, 잘못된 입력값을 도메인까지 끌고와서 유효성 검사를 하는게 과연 옳은일인가? 싶은 생각이 들었습니다. 그래서 만약에 로또 당첨 번호에 대한 입력을 받는다고 하면, 모든 검사를 도메인에서 하기보다는

* 컨트롤러에서 유효성 검사를 해줘야할 부분: 숫자가 아닌 입력이 들어왔는지
* 로또, 로또 번호에서 유효성 검사를 해줘야할 부분: 1-45의 숫자가 아닌지, 6개의 숫자인지, 중복된 숫자가 있는지

이런식으로 유효성 검사에 대한 책임도 분리를 시켰고 그에따라 필요 없어진 생성자는 삭제하였습니다. 괜찮은 방법인지 궁금합니다!

리뷰 항상 감사드립니다😊

toneyparky commented on src/main/java/lotto/domain/Profit.java:
```diff
@@ -0,0 +1,14 @@
+
+    private Money profit;
+
+    public Profit(Money profit) {
+        this.profit = profit;
+    }
+
+    public double calculateProfitRate(Money purchaseAmount) {
+        return (double)profit.getMoney() / purchaseAmount.getMoney();
+    }
```
리팩터링 이후에는 `Money` 객체를 필드로 가지는 클래스가 되었군요! 

하는 행위는 `Money` 객체로부터 돈을 가져와서 나눠주는 역할을 하구요! 

이정도는 `Money` 객체만 가지고 처리해줘도 되려나 싶지만 마루가 수익이라는 의도를 더 표현하고 싶으시다고 이해했어요!

toneyparky commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,52 @@
+        return this.isBonus;
+    }
+
+    public static List<Rank> getSortedRanks() {
+        return new ArrayList<>(Arrays.stream(values())
+            .sorted(Comparator.comparing(Rank::getPrizeMoney))
+            .filter(rank -> rank != NOTHING)
+            .collect(Collectors.toList()));
+    }
```
8ca280f 요 커밋에서 진행한 리팩터링이네요~!

어떤 의도인지 커밋 메시지에 적어주신 외부에서 내부 값을 바꾸지 못하도록 수정했다는게 불변을 의도하신걸까요?

답변주신다면 제가 배워갈 수 있을거 같네요 😉

chawani replied:
네 맞습니다! 불변을 위해서 새 ArrayList로 만들어 리턴해주었습니다

toneyparky replied:
그렇군요! 불변의 의미가 리스트에 값을 더하거나 제외할 수 없음이겠죠?

그렇다면 제가 알고 있는바로는 새 ArrayList로 묶어서 반환하더라도 불변을 유지하지 않을거에요. 

해당 함수의 반환값에 add를 해볼까요? 혹은 테스트 코드를 만들어서 확인해보셔도 좋겠네요!

불변 리스트를 반환하고 싶다면 `Collections.unmodifiableList()`를 활용해보셔도 좋을 것 같아요😀

toneyparky commented on src/test/java/lotto/domain/LottoTest.java:
```diff
@@ -0,0 +1,58 @@
+
+        assertEquals(Rank.SECOND, lotto.getRank(new WinningLotto(winLotto, bonusBall)));
+    }
+
```
불필요한 공백이 있네요~!

toneyparky commented on src/test/java/lotto/domain/MoneyTest.java:
```diff
@@ -0,0 +1,36 @@
+        Money money = new Money(0);
+        money.add(100);
+
+        assertTrue(money.getMoney() == 100);
```
보통 `assertTrue`는 내부에서 비교해서 true인지를 검사하기보다는 검사하고자 하는 값이 boolean일 때에 사용해요. 

이럴 경우는, 사용하신 `junit.jupiter` 패키지의 `assertEquals`나 `org.assertj.core.api.Assertions.assertThat`을 이용하여 `assertThat().isEqualTo()` 처럼 사용하길 권장드립니다.

자바의 테스트 검증에 있어 junit의 jupiter와 assertj의 차이를 공부해보셔도 재미있을거에요.

toneyparky commented on src/test/java/lotto/domain/RankTest.java:
```diff
@@ -0,0 +1,17 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+import java.util.List;
```
사용하지 않는 임포트가 있네요~!

toneyparky commented:
마루가 질문으로 적어주신 유효성 검사를 분리한다는 내용 좋아요!

실제로 유효성 검사의 책임을 모두 도메인에만 있지 않기에 적절하게 배분하셨다고 생각해요. 

우테코 레벨 2에서 스프링을 학습하실텐데 그 때에는 사용자로부터 요청받는 정보에 대한 유효성 검사와 내부의 도메인 생성에 대한 유효성 검사를 각각 진행해야 하는 경우가 있기에 염두해두시면 좋을 것 같아요. 추가로 말씀드리면 유효성 검사의 종류를 딱 잘라서 너는 이거 너는 저거라고 나뉘지는 않아요. 각 레이어(지금은 컨트롤러와 모델)에서 필요한 유효성 검사가 있으면 해주면 되는거에요. 이러다가 중복도 발생하게 되구요!

좋은 리팩토링을 진행하셨다고 생각합니다!

toneyparky reviewed (approved):
리팩터링 반영하느라 고생 많으셨어요! 코드가 훨씬 통일감 있어졌네요!

그리고 컨트롤러가 서술적으로 바뀐거 같아요. 컨트롤러만 보고도 기존보다는 로직을 파악하기 쉬워졌습니다 👍

기존에 남겼던 요 질문에 대해서 넘어가시기보다 댓글로 남겨주시면 마루의 학습에도 도움이 될거에요.
https://github.com/woowacourse/java-lotto/pull/386#discussion_r813924273

리뷰와 질문을 몇가지 남겼지만 문제가 있는 사항은 아니라 1 단계 미션은 머지해도 괜찮겠어요!
질문에 대해서는 댓글로 답변해주시고 반영은 2 단계 미션에서 진행해보죠!

고생하셨습니다~!
